lib/SPIRV/SPIRVLowerMemmove.cpp

```diff
@@ -76,21 +76,30 @@ class SPIRVLowerMemmoveBase {
     Value *NumElements = nullptr;
     uint64_t ElementsCount = 1;
     if (SrcTy->isArrayTy()) {
-      NumElements = Builder.getInt32(SrcTy->getArrayNumElements());
       ElementsCount = SrcTy->getArrayNumElements();
+      NumElements = Builder.getInt32(ElementsCount);
     }
-    if (((ElementsCount > 1) && (Mod->getDataLayout().getTypeSizeInBits(
-                                     SrcTy->getPointerElementType()) *
-                                     ElementsCount !=
-                                 Length->getZExtValue() * 8)) ||
-        ((ElementsCount == 1) &&
-         (Mod->getDataLayout().getTypeSizeInBits(
-              SrcTy->getPointerElementType()) < Length->getZExtValue() * 8)))
-      report_fatal_error("Size of the memcpy should match the allocated memory",
+    // Check whether Src layout size matches the length parameter of
+    // llvm.memmove intrinsic. It is possible that for a single element Src
+    // type size doesn't match memory size to move (for example we copy N
+    // elements of one integer array to an address of a first element of
+    // another integer array). So we ignore this check here and handle it
+    // by setting alloca type instruction.
+    const uint64_t LenBits = Length->getZExtValue() * 8;
+    const uint64_t LayoutTypeBits =
+        Mod->getDataLayout().getTypeSizeInBits(SrcTy->getPointerElementType()) *
```

Comment:
`SrcTy` is not necessarily the original type used in `llvm.memmove`, because we went through casts at lines 71-72.

Reply:
Yes and I believe it's done by design, for example we can call:
call void @llvm.memmove.p1i8.p1i8.i32(i8 * %dsr, i8 * %src, i32 4, i1 false)
where src and dst pointer are of i8 type whilst the real value is of i32 type (it was just bitcasted away). In this case we shall allocate not a vector [i8 x 4] but a single i32 chunk of memory for a temporary variable.